File: src/LeetCode/MoveZeroes.java
package LeetCode;

public class MoveZeroes {
    final private int[] nums;

    public MoveZeroes(int[] nums) {
        this.nums = nums;
    }

    public void swap(int[] nums, int leftIndex, int rightIndex) {
        int tmp = nums[leftIndex];
        nums[leftIndex] = nums[rightIndex];
        nums[rightIndex] = tmp;
    }

    public int[] naive() {
        int leftIndex = 0;
        int rightIndex = 1;
        int n = nums.length;

        while (rightIndex < n) {
            int leftNum = nums[leftIndex];
            int rightNum = nums[rightIndex];

            if (leftNum != 0) {
                leftIndex++;
                rightIndex++;
            } else if (rightNum == 0) {
                rightIndex++;
            } else {
                swap(nums, leftIndex++, rightIndex++);
            }

        }


        return nums;
    }


}

File: dsaLeetCode/arrayAndString/moveZeroes.js
// Space Complexity: O(n)
// Time Complexity: O(1)
const fn = (nums) => {
  let l = 0;
  let r = 0;
  const n = nums.length;

  while(r < n) {
    const leftNum = nums[l];
    const rightNum = nums[r];

    if(leftNum !== 0) {
      l++;
    } else if(rightNum !== 0) {
      // switch
      [nums[l], nums[r]] = [nums[r], nums[l]];
      l++;
    };

    r++;
  };

  return nums;
};

const nums = [0, 0,0, 0, 0];

console.log(fn(nums));
